### PR TITLE
Update ViennaRNA source files (2.4.17)

### PR DESCRIPTION
**Pull request by sjanssen2**
This is a bigger PR, which shall update files regarding RNA energy computation by approx. 10 years. Those files originally stem from the Vienna RNA package (probably some version number prior to their 2.x.x). The rnalib.c code is an adapter between the ViennaRNA world and Bellman's GAP and we herewith intend to apply bugs and enhancements made in the past years in the Vienna package and support future updates by a more clear structure.

Here is some information about the introduced changes:
 - All files in `librna/ViennaRNA` are identical copied from https://github.com/ViennaRNA/ViennaRNA/tree/v2.4.17/src/ViennaRNA, i.e. ignore them in your review!
 - the new source files need some specific gcc compile flags. These are set via the file `librna/config.h` which is generated by the `./configure` script. This script is itself product of a generation process, namely https://de.wikipedia.org/wiki/GNU_Build_System, i.e. ignore `configure`, `librna/config.h.in` and `librna/configure`, but carefully inspect `configure.ac` and `librna/configure.ac` as they are the input for the generation process. Some more details are documented in `documentation/developer_notes`
 - the `Makefile` need some changes to reflect additional objects and new subdirectory name
 - `librna/rnalib.c` has to address slightly different function and data structure names
 - `makefiles/run-librna.sh` and `makefiles/run.sh` are used for `make install` and need to copy parameter files from a novel dedicated sub-directory `librna/paramfiles/*`
 - same as above is true for tests, i.e. `testdata/paraltest/config` and `testdata/regresstest/config`

### CODE REVIEW

**Comment by sjanssen2** [comment]
**This issue was solved via great help of @RaumZeit. Many thanks!**

However, I get a seg fault when passing a parameter files to the compiled binary. I figure this might be related to wrong/missing compiler flags when creating the shared library `librna.so`, since this error cannot be reproduced when debugging in Eclipse.

Here is a bt of the failing core dump:
```
(base) t490s x86_64 /Daten/Git/jlab/fold-grammars/Misc/Applications/RNAshapes/x86_64-linux-gnu>coredumpctl gdb -1
           PID: 29520 (RNAshapes_eval_)
           UID: 1000 (sjanssen)
           GID: 1000 (sjanssen)
        Signal: 11 (SEGV)
     Timestamp: Fri 2021-04-16 13:07:07 CEST (4s ago)
  Command Line: ./RNAshapes_eval_macrostate -T 57 -P /home/sjanssen/src/ViennaRNA-2.4.17/misc/rna_turner2004.par CCaCCaaaGGCCaaaGGaGG ((.((...))((...)).))
    Executable: /Daten/Git/jlab/fold-grammars/Misc/Applications/RNAshapes/x86_64-linux-gnu/RNAshapes_eval_macrostate
 Control Group: /user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
          Unit: user@1000.service
     User Unit: gnome-terminal-server.service
         Slice: user-1000.slice
     Owner UID: 1000 (sjanssen)
       Boot ID: de9b7b3d57674135bc9badcff29e8d32
    Machine ID: f3d85c83fd6e458aba76dbf56f683032
      Hostname: t490s
       Storage: /var/lib/systemd/coredump/core.RNAshapes_eval_.1000.de9b7b3d57674135bc9badcff29e8d32.29520.1618571227000000.lz4
       Message: Process 29520 (RNAshapes_eval_) of user 1000 dumped core.
                
                Stack trace of thread 29520:
                #0  0x00007f1c7c7a4a6d __GI___libc_free (libc.so.6)
                #1  0x00007f1c7d0af015 n/a (/home/sjanssen/lib/librna.so)

GNU gdb (GDB) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /Daten/Git/jlab/fold-grammars/Misc/Applications/RNAshapes/x86_64-linux-gnu/RNAshapes_eval_macrostate...
(No debugging symbols found in /Daten/Git/jlab/fold-grammars/Misc/Applications/RNAshapes/x86_64-linux-gnu/RNAshapes_eval_macrostate)
[New LWP 29520]
Core was generated by `./RNAshapes_eval_macrostate -T 57 -P /home/sjanssen/src/ViennaRNA-2.4.17/misc/r'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f1c7c7a4a6d in free () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) bt
#0  0x00007f1c7c7a4a6d in free () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f1c7d0af015 in vrna_params_load () from /home/sjanssen/lib/librna.so
#2  0x00007f1c7d0a54b5 in librna_read_param_file () from /home/sjanssen/lib/librna.so
#3  0x000056149a331c7a in gapc::Opts::parse(int, char**) ()
#4  0x000056149a32b0ab in main ()
(gdb) q
```

**Review by kmaibach — APPROVED** [review]

File: librna/rnalib.c [hunk 1]
```diff
@@ -17,18 +17,18 @@
 
 #include "rnalib.h"
 
-#include "vienna/data_structures.h"
-static paramT  *P = 0;
+#include "ViennaRNA/datastructures/basic.h"
+static vrna_param_t  *P = 0;
 // if nonzero use logarithmic ML energy in energy_of_struct
 // used in vienna/fold_vars.c, set_model_details()
```
[the remark this answers is not in the document]
Comment (sjanssen2) [reply librna/rnalib.c]: very thorough :-) Thanks!
Actually, `#include` is no comment but "importing" functions from header (source) files. The paths are relative to working dir, i.e. you can specify where to look for them via `gcc -I xxx`.
You are correct with line 1022. I fixed it.

[the remark this answers is not in the document]
Comment (kmaibach) [reply librna/rnalib.c]: Oh, I didn't mean the `#include`, I meant line 23 `// used in vienna/fold_vars.c, set_model_details()`

[the remark this answers is not in the document]
Comment (sjanssen2) [reply librna/rnalib.c]: ups. Sorry. This comment and the according code can be deleted alltogether

**Pull request merged**